blosc/shuffle-sse2.c

```diff
@@ -66,63 +66,65 @@ shuffle2_sse2(uint8_t* dest, const uint8_t* src, size_t size)
       xmm1[k*2+1] = _mm_unpackhi_epi64(xmm0[k], xmm0[k+1]);
     }
     /* Store the result vectors */
+    uint8_t* const dest_for_jth_element = dest + j;
     for (k = 0; k < 2; k++) {
-      ((__m128i *)dest)[k*numof16belem+i] = xmm1[k];
+      _mm_storeu_si128((__m128i*)(dest_for_jth_element + (k * total_elements)), xmm1[k]);
     }
   }
 }
 
-
 /* Routine optimized for shuffling a buffer for a type size of 4 bytes. */
 static void
-shuffle4_sse2(uint8_t* dest, const uint8_t* src, size_t size)
+shuffle4_sse2(uint8_t* dest, const uint8_t* src,
+  const size_t vectorizable_elements, const size_t total_elements)
 {
-  size_t i, j, k;
-  size_t numof16belem;
+  static const size_t bytesoftype = 4;
+  size_t i;
+  int j;
   __m128i xmm0[4], xmm1[4];
 
-  numof16belem = size / (16*4);
-  for (i = 0, j = 0; i < numof16belem; i++, j += 16*4) {
-    /* Fetch and transpose bytes and words in groups of 64 bytes */
-    for (k = 0; k < 4; k++) {
-      xmm0[k] = _mm_loadu_si128((__m128i*)(src+j+k*16));
-      xmm1[k] = _mm_shuffle_epi32(xmm0[k], 0xd8);
-      xmm0[k] = _mm_shuffle_epi32(xmm0[k], 0x8d);
-      xmm0[k] = _mm_unpacklo_epi8(xmm1[k], xmm0[k]);
-      xmm1[k] = _mm_shuffle_epi32(xmm0[k], 0x04e);
-      xmm0[k] = _mm_unpacklo_epi16(xmm0[k], xmm1[k]);
+  for (i = 0; i < vectorizable_elements; i += sizeof(__m128i)) {
+    /* Fetch 16 elements (64 bytes) then transpose bytes and words. */
+    for (j = 0; j < 4; j++) {
+      xmm0[j] = _mm_loadu_si128((__m128i*)(src + (i * bytesoftype) + (j * sizeof(__m128i))));
+      xmm1[j] = _mm_shuffle_epi32(xmm0[j], 0xd8);
+      xmm0[j] = _mm_shuffle_epi32(xmm0[j], 0x8d);
+      xmm0[j] = _mm_unpacklo_epi8(xmm1[j], xmm0[j]);
+      xmm1[j] = _mm_shuffle_epi32(xmm0[j], 0x04e);
+      xmm0[j] = _mm_unpacklo_epi16(xmm0[j], xmm1[j]);
     }
     /* Transpose double words */
-    for (k = 0; k < 2; k++) {
-      xmm1[k*2] = _mm_unpacklo_epi32(xmm0[k*2], xmm0[k*2+1]);
-      xmm1[k*2+1] = _mm_unpackhi_epi32(xmm0[k*2], xmm0[k*2+1]);
+    for (j = 0; j < 2; j++) {
+      xmm1[j*2] = _mm_unpacklo_epi32(xmm0[j*2], xmm0[j*2+1]);
+      xmm1[j*2+1] = _mm_unpackhi_epi32(xmm0[j*2], xmm0[j*2+1]);
     }
     /* Transpose quad words */
-    for (k = 0; k < 2; k++) {
-      xmm0[k*2] = _mm_unpacklo_epi64(xmm1[k], xmm1[k+2]);
-      xmm0[k*2+1] = _mm_unpackhi_epi64(xmm1[k], xmm1[k+2]);
+    for (j = 0; j < 2; j++) {
+      xmm0[j*2] = _mm_unpacklo_epi64(xmm1[j], xmm1[j+2]);
+      xmm0[j*2+1] = _mm_unpackhi_epi64(xmm1[j], xmm1[j+2]);
     }
     /* Store the result vectors */
-    for (k = 0; k < 4; k++) {
-      ((__m128i *)dest)[k*numof16belem+i] = xmm0[k];
+    uint8_t* const dest_for_ith_element = dest + i;
+    for (j = 0; j < 4; j++) {
+      _mm_storeu_si128((__m128i*)(dest_for_ith_element + (j * total_elements)), xmm0[j]);
```

Comment:
Well, I just spent a bit more time on this and redid my benchmarks and I am not seeing the slow-down anymore, or a very small difference (< 1%) (using a i5-3380M @ 2.90GHz).